driver-core/src/main/com/mongodb/internal/operation/AsyncOperationHelper.java

```diff
@@ -138,6 +138,9 @@ static <R, T extends ReferenceCounted> void withAsyncSuppliedResource(final Asyn
                     AsyncCallbackSupplier<R> curriedFunction = c -> function.apply(resource, c);
                     curriedFunction.whenComplete(resource::release).get(errorHandlingCallback);
                 } catch (Exception e) {
+                    if (resource.getCount() > 0) {
```

Comment:
This more closely mirrors the SyncOperationHelper and will ensure any errors thrown outside of the callback also release the resource.